soh/soh/Enhancements/randomizer/randomizer.cpp

```diff
@@ -3921,20 +3924,26 @@ void DrawRandoEditor(bool& open) {
                     "\n"
                     "Off - No dungeons will have their keys replaced with keyrings.\n"
                     "\n"
-                    "Random - A random amount of dungeons(0-8) will have their keys replaced with keyrings.\n"
+                    "Random - A random amount of dungeons(0-9) will have their keys replaced with keyrings.\n"
                     "\n"
                     "Count - A specified amount of randomly selected dungeons will have their keys replaced with keyrings.\n"
                     "\n"
-                    "Selection - Hand select which dungeons will have their keys replaced with keyrings."
+                    "Selection - Hand select which dungeons will have their keys replaced with keyrings.\n"
+                    "\n"
+                    "Selecting key ring for dungeons will have no effect if Small Keys are set to Start With or Vanilla.\n"
+                    "\n"
+                    "Selecting key ring for Thieves' Hideout will have no effect if Thieves' Hideout keys are in vanilla "
```

Comment:
I think this comment here is why this didn’t go in with the rest, I don’t think I was sure how to properly handle this

Reply:
yeah, i feel like we probably want to expose this more clearly than just having a tooltip here. right now, if a player picks count 3, there will definitely be 3 keyrings in the pool. adding this into the mix makes it so there may only be 2 depending on their GF key settings.

i think to implement this we should make sure the UI communicates the setting interaction (should probably do a similar thing for start with and vanilla dungeon keys, but that feels less likely to confuse people)

Reply:
@briaguya-ai I'm not completely opposed to changing the UI, but am weary of the changes needed to track all that both in the randomizer menu, and in the 3drando code.

Before I go that approach, I'm curious if adjusting the tooltip description would be sufficient to account for any oddities. Here is one example I came up
> Count - Up to a specified amount of randomly selected dungeons will have their keys replaced with keyrings. Exact count depends on key settings.

Reply:
As far as I can tell from looking at the 3drando code, Gerudo Fortress is not included in the dungeons that can be randomly selected from the count to have keyrings. The code that picks the random dungeons to give key rings to is around `settings.cpp L2969` pulls from `dungeonList`, filters out dungeons with no keys, shuffles that list, and then gives key rings to the first `keyRingCount` dungeons in the shuffled list. Said `dungeonList` does not have Gerudo Fortress.

Meanwhile whether or not to add a Gerudo's Fortress keying to the pool is handled separately from everything else in `item_pool.cpp L828-872`. So whether it has a key ring or not is handled entirely independently of all the other dungeons, and it having or not having a key ring will not affect the key ring count of the other dungeons.

Therefore, the only thing to actually track for whether or not to show this option is what Gerudo Fortress access is set to. We should hide this option if it is set to either Open, Vanilla, or Fast, as those options either remove Gerudo Fortress keys altogether, or they require the Gerudo Fortress keys to be in their vanilla locations. Otherwise we show the keyring option.

Reply:
So TL;DR

@briaguya-ai, the key ring count would not be affected by the possibility of GF getting randomly selected and not being able to have a key ring, because it actually can't be randomly selected at all, see above.

@Archez The logic for hiding or showing the GF Key Ring option should be pretty simple and only dependent on GF access settings, see above.

Reply:
Oh wait, I made a mistake here, I had only compared our current code, not the other changes you made in 3drando/settings.cpp which are also coming from upstream 3drando. Ignore the above for now.

Reply:
That being said, it appears that 3drando removed the ability to have a certain amount of keyrings at the same time as they added the ability to have Gerudo Fortress keyrings. Probably for similar reasons that we are having this debate right now.

Reply:
Can ImGui handle the max number of the slider changing at runtime? If so, I think it shouldn't be too hard to actually just decrease the max count to 8 if we get in a situation where Gerudo Fortress can't have a KeyRing (i.e. if GerudoFortress is at anything other than Normal) and also hide the checkbox for Gerudo Fortress key ring. Then in 3drando we can start with keyRingOptions not having RingFortress in it, and pushing it onto that vector before setting key rings on or off according to the keyRingCount if we can have a GerudoFortressKeyring

Reply:
> Can ImGui handle the max number of the slider changing at runtime

IIRC yes

Reply:
I PR'ed a fix to the Archez's branch that adds this extra validation to disable the check box and change the max key ring count. https://github.com/Archez/Shipwright/pull/2

Reply:
Thanks for the PR, assuming this will resolve our concerns related to the wording now 👍

Reply:
Provided everyone else is fine with how I did it, then yes.